interfaces/builtin/qualcomm_ipc_router.go

```diff
@@ -40,32 +45,214 @@ const qipcrtrBaseDeclarationSlots = `
 capability net_admin,
+capability net_bind_service,
+
+unix (bind, listen) type=seqpacket addr="###SOCKET_ADDRESS###",
+`
+
+const qipcrtrConnectedSlotAppArmor = `
+unix (accept, send, receive) type=seqpacket addr="###SOCKET_ADDRESS###" peer=(label=###PLUG_SECURITY_TAGS###),
+`
+
+const qipcrtrConnectedPlugSecCompCompat = `
+# Description: allows access to the Qualcomm IPC Router sockets
+bind
+
+# We allow AF_QIPCRTR in the default template since it is mediated via the AppArmor rule
+#socket AF_QIPCRTR
 `
 
-const qipcrtrConnectedPlugSecComp = `
+const qipcrtrPermanentSlotSecComp = `
 # Description: allows access to the Qualcomm IPC Router sockets
 bind
+accept
+accept4
+listen
 
 # We allow AF_QIPCRTR in the default template since it is mediated via the AppArmor rule
 #socket AF_QIPCRTR
 `
 
+const (
+	// qcipcAttrib is a label that must match between plug and slot
+	qcipcAttrib = "qcipc"
+	// addressAttrib is the allowed socket address and is specified in the slot
+	addressAttrib = "address"
+)
+
 type qualcomIPCRouterInterface struct {
 	commonInterface
 }
 
-func (iface *qualcomIPCRouterInterface) BeforeConnectPlug(plug *interfaces.ConnectedPlug) error {
+func (iface *qualcomIPCRouterInterface) Name() string {
+	return "qualcomm-ipc-router"
+}
+
+func (iface *qualcomIPCRouterInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              qipcrtrSummary,
+		BaseDeclarationSlots: qipcrtrBaseDeclarationSlots,
+		// backward compatibility
+		ImplicitOnCore: true,
+		// backward compatibility
+		ImplicitOnClassic: true,
+	}
+}
+
+func setSnippetSocketAddress(slot *interfaces.ConnectedSlot, snippet string) (string, error) {
+	var address string
+	if err := slot.Attr(addressAttrib, &address); err != nil {
+		return "", err
+	}
+
+	return strings.ReplaceAll(snippet, "###SOCKET_ADDRESS###", address), nil
+}
+
+// Note that we cannot use implicitSystemConnectedSlot as legacy slots can be
+// for UC too in this case.
+func isConnectedSlotSystem(slot *interfaces.ConnectedSlot) bool {
+	if slot.Snap().Type() == snap.TypeOS || slot.Snap().Type() == snap.TypeSnapd {
+		return true
+	}
+	return false
+}
+
+func isSlotInfoSystem(slot *snap.SlotInfo) bool {
+	if slot.Snap.SnapType == snap.TypeOS || slot.Snap.SnapType == snap.TypeSnapd {
+		return true
+	}
+	return false
+}
+
+func validateLabel(qcipc string) error {
+	if len(qcipc) == 0 {
+		return fmt.Errorf("qualcomm-ipc-router %s attribute cannot be empty", qcipcAttrib)
+	}
+	if err := naming.ValidateLabel(qcipc); err != nil {
+		return fmt.Errorf("bad name for %s attribute: %v", qcipcAttrib, err)
+	}
+	return nil
+}
+
+func validateAddress(address string) error {
+	if len(address) == 0 {
+		return fmt.Errorf("qualcomm-ipc-router %s attribute cannot be empty", qcipcAttrib)
+	}
+	// same as from ValidateNoAppArmorRegexp, but with globbing
+	const aareWithoutGlob = `?[]{}^"` + "\x00"
```

Comment:
do we really want to allow for globbing?

Reply:
I think some flexibility with that would be fine as maybe socket addresses are not always fixed, but I'll take the advice from @alexmurray about this.

Reply:
We don't normally allow to use AppArmor globbing etc in slot/plug attributes - and so unless we have a good use-case for it, I think we are safer to stick to having explicit values only and so should use `ValidateNoAppArmorRegexp`.